Libraries/Connect/Internal/Streaming/ClientOnlyStream.swift

```diff
@@ -0,0 +1,80 @@
+// Copyright 2022-2024 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import SwiftProtobuf
+
+/// Concrete implementation of `ClientOnlyStreamInterface`.
+///
+/// The complexity around configuring callbacks on this type is an artifact of the library
+/// supporting both callbacks and async/await. This is internal to the package, and not public.
+final class ClientOnlyStream<Input: ProtobufMessage, Output: ProtobufMessage>: @unchecked Sendable {
+    private let onResult: @Sendable (StreamResult<Output>) -> Void
+    private let receivedMessageCount = Locked(0)
+    /// Callbacks used to send outbound data and close the stream.
+    /// Optional because these callbacks are not available until the stream is initialized.
+    private var requestCallbacks: RequestCallbacks<Input>?
+
+    private struct NotConfiguredForSendingError: Swift.Error {}
+
+    init(onResult: @escaping @Sendable (StreamResult<Output>) -> Void) {
+        self.onResult = onResult
+    }
+
+    /// Enable sending data over this stream by providing a set of request callbacks to route data
+    /// to the network client. Must be called before calling `send()`.
+    ///
+    /// - parameter requestCallbacks: Callbacks to use for sending request data and closing the
+    ///                               stream.
+    ///
+    /// - returns: This instance of the stream (useful for chaining).
+    @discardableResult
+    func configureForSending(with requestCallbacks: RequestCallbacks<Input>) -> Self {
+        self.requestCallbacks = requestCallbacks
+        return self
+    }
+
+    /// Send a result to the consumer after doing additional validations for client-only streams.
+    /// Should be called by the protocol client when a result is received.
+    ///
+    /// - parameter result: The new result that was received.
+    func receive(_ result: StreamResult<Output>) {
+        let receivedMessageCount = self.receivedMessageCount.perform { value in
+            if case .message = result {
+                value += 1
+            }
+            return value
+        }
```

Comment:
I'm not really following this block. I _think_ it means that `receive` must be called 2x in order to see the count bump to 2. Is that right? If so, what if the caller only calls `receive` once? It's not clear to me how we are actually validating that there isn't more than one message in the response, other than returning an error if there is when the application asks for another.

Reply:
This is related to the wiring I alluded to in my earlier comments - `receive` is an internal function called by the protocol client to pass a network chunk to this handle. It's not something called by end-users, it's called by internals _to pass that result on to end users_. I'll rename these to try to provide a bit more clarity (and I'd really like to remove some of this in the future if we can centralize on async/await at some point). Let me know if that clears this up